### PR TITLE
[react-redux] Configurable RootState with module augmentation

### PR DESCRIPTION
**Pull request by Lazyuki**
With this PR, users can use Typescript's [module augmentation](https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation) to specify the type of `RootState` in which all the `react-redux` functions can infer from. (I got this idea from how `styled-components` does `DefaultTheme` ) https://github.com/DefinitelyTyped/DefinitelyTyped/blob/7cb38cbc6d4f3092270331f90c16d29d1fc43cc2/types/styled-components/index.d.ts#L410-L417
All users need to do is create a `d.ts` file that extends this `RootState` with their own state. This works particularly well with redux since you are supposed to create only one "root state" and all the components share this state
e.g.
```tsx
// my-redux.d.ts
import 'react-redux';
type UserState = { name: string };
type CacheState = { favorites: string[] };

type MyState = { user: UserState, cache: CacheState };

declare module 'react-redux' {
  export interface ConfigurableRootState extends MyState {}
}
```

Then, anywhere in their code, they could simply do 
```tsx
connect(
    ({ user, cache }) => (
        { 
            username: user.name,
            favorites: cache.favorites,
        }
    ))(Component);
```
without manually specifying types.
The root state type would be automatically inferred in other functions like
```tsx
const username = useSelector(state => state.user.name);
```
as well. 

I've tested this in my own project and it works great, but I may have overlooked unintended side effects. Let me know if this is something that you guys tried but decided against it. 

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.) **Can't test module augmentation in those test files**
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <see styled-components link>

### CODE REVIEW

**Comment by andrewbranch** [comment]
I’d like to wait for feedback from the package owners on this. Thanks for your patience with the holidays. In the meantime, it would be advisable to add a test for this functionality.

**Comment by Lazyuki** [comment]
@andrewbranch thanks for the update. However if I understand it correctly, there's no way to test module augmentation feature at least within DefinitelyTyped? If I augment this module anywhere in the test wouldn't it affect all the other tests?

**Comment by andrewbranch** [comment]
Hm yeah, I think that’s true. I think I’d personally rather see the tests have an augmentation than not, but I guess it’s debatable. I don’t think it’s a blocker for merging. I’ll leave it up to you and the package owners.

**Comment by Lazyuki** [comment]
@andrewbranch Would I need to do something to get attention of the "package owners"? I don't know who those would be.

**Comment by andrewbranch** [comment]
TypeScript bot has already pinged them all: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/41031#issuecomment-565658528

Normally after two weeks I would start bugging people more or just go ahead and merge it since it looks reasonable to me, but with the holidays I’m willing to give it until the end of next week.

**Comment by Lazyuki** [comment]
@tkqubo @kenzierocks @clayne11 @tansongyang @NicholasBoll @mDibyo @Kallikrein @val1984 @jrakotoharisoa @apapirovski @surgeboris @soerenbf @MrWolfZ @DylanVann 
bump if you missed this during the holidays

File: types/react-redux/index.d.ts
Comment (MrWolfZ) [inline types/react-redux/index.d.ts]: I'm not sure I agree with this change. By doing this the type of `TState` becomes `any` when no type parameters are given, while before this would result in `unknown` which seems more reasonable to me

File: types/react-redux/index.d.ts
Comment (MrWolfZ) [inline types/react-redux/index.d.ts]: Making `SafeRootState` `any` instead of `{}` as a default can make certain errors harder to spot

**Review by MrWolfZ — COMMENTED** [review]
As far as I can see this change seems reasonable enough and should be "safe" in the sense that there is no breaking change for existing projects. However, certain usages of the types become more permissive by using `any` instead of `{}` as a default. Not sure if that is desirable. Since I do not use react-redux in any of my projects I have a hard time to judge how big the benefit of this is. It's probably also quite difficult to know about this feature without looking at the types themselves, so maybe we'd need some hints in the comments of some of the other types that mention this feature.

File: types/react-redux/index.d.ts
Comment (Lazyuki) [reply types/react-redux/index.d.ts]: Good point, I'll fix it

**Comment by Lazyuki** [comment]
Hey @MrWolfZ I've updated some stuff. Can you review again?

File: types/react-redux/index.d.ts
Comment (MrWolfZ) [inline types/react-redux/index.d.ts]: Isn't this redundant? Make it an empty object if it is an empty object?

File: types/react-redux/index.d.ts
Comment (MrWolfZ) [inline types/react-redux/index.d.ts]: This is a breaking change now. I think you should keep all other places as `{}` while keeping this `any`

File: types/react-redux/react-redux-tests.tsx
Comment (MrWolfZ) [inline types/react-redux/react-redux-tests.tsx]: this test change shows that this is a breaking change. please ensure this test still works without any changes to it

File: types/react-redux/react-redux-tests.tsx
Comment (MrWolfZ) [inline types/react-redux/react-redux-tests.tsx]: this test change shows that this is a breaking change. please ensure this test still works without any changes to it

**Review by MrWolfZ — CHANGES_REQUESTED** [review]
As written in the code comments, please ensure that your changes do not contain any breaking changes

File: types/react-redux/index.d.ts
Comment (Lazyuki) [reply types/react-redux/index.d.ts]: Yeah this is a breaking change but I'm not sure why it's `any` in the first place. I feel like it should have been an empty object.

File: types/react-redux/index.d.ts
Comment (Lazyuki) [reply types/react-redux/index.d.ts]: Oops, you are right. This can be removed

File: types/react-redux/index.d.ts
Comment (MrWolfZ) [reply types/react-redux/index.d.ts]: That may be, but unfortunately there are already too many apps out there that depend on this. Therefore, we need to revert this change.

**Review by MrWolfZ — APPROVED** [review]

**Comment by oidualc** [comment]
This change broke the inline function definition on useSelector on JavaScript files when TypeScript check is enabled (e.g. on VSCode).
For example this expression `useSelector(state => state.myproperty)` now gives the error `Property 'myproperty' does not exist on type 'DefaultRootState'. ts(2339)` because now `state`is of type `DefaultRootState`.
Can this pull request be rethought taking this use case into account?

**Pull request merged**
